billing_scanner/scanner.py

```
@@ -0,0 +1,201 @@
+        """
+        logger.info(f"Processing log file: {key}")
+        content = self.download_log_file(key)
+        if not content:
+            return False
+
+        # Aggregate log lines by a composite key: (log filename, workspace, SKU)
+        aggregation = defaultdict(
+            lambda: {
+                "data_size": 0,
+                "earliest": None,
+                "latest": None,
+                "workspace": None,
+                "sku": None,
+            }
+        )
+        for line in content.splitlines():
+            event_data = self.process_log_line(line, key)
+            if event_data:
+                group_key = event_data["aggregation_key"]
+                group = aggregation[group_key]
+                group["data_size"] += event_data["data_size"]
+                ts = datetime.fromisoformat(event_data["timestamp"].replace("Z", ""))
+                if group["earliest"] is None or ts < group["earliest"]:
+                    group["earliest"] = ts
+                if group["latest"] is None or ts > group["latest"]:
+                    group["latest"] = ts
+                group["workspace"] = event_data["workspace"]
+                group["sku"] = event_data["sku"]
+
+        # For each aggregated group, generate one BillingEvent.
+        for group_key, group in aggregation.items():
+            # Generate a unique and repeatable UUID using uuid.uuid5 from the composite key.
+            event_uuid = uuid.uuid5(uuid.NAMESPACE_DNS, f"{group_key}")
+            billing_event = messages.BillingEvent()
+            billing_event.uuid = str(event_uuid)
+            billing_event.event_start = group["earliest"].isoformat() + "Z"
+            billing_event.event_end = group["latest"].isoformat() + "Z"
+            billing_event.sku = group["sku"]
+            billing_event.workspace = group["workspace"]
+            billing_event.quantity = float(group["data_size"])
+            self.publish_event(
```

This appears to ignore errors when publishing the billing event to the topic, then go right ahead and mark the file as processed in the recorded state.

---

I have added try and catch block to handle the errors